### ISSUE TITLE
ICS 4: Timeout Height

### ISSUE THREAD

**Issue by mdyring** [cosmos/ibc#353]
Disclaimer: I haven't had an opportunity to fully dive into the ICS standards yet.

Taking a glance at ICS 4, I noticed that the concept of a "Timeout Height" is used extensively.

As any blockchain has a monotonically increasing height, using height for timeouts seems like a no-brainer (at least initially :-D).

I'd like to offer some thoughts on why "block height" is not well suited for for timeouts:
* The natural unit for a "time out" is a point in time
* Tendermint based chains can be configured for on-demand block generation (as we are doing with https://github.com/e-money/em-ledger), I expect this to be more commonplace over time.

Specifically, my fear is that chains with on-demand block generation will be more "bursty" in nature, i.e. a large amount of blocks produced in a short timespan followed by periods of little activity. As the pattern is unpredictable, it makes it hard to reason what a single timeout height should be.

Since Tendermint has a concept of block time, I imagine that could be used instead. But that will most likely make it harder to implement ICS 4 outside of Tendermint.

Any thoughts on this?

**Comment by cwgoes** [comment]
The concern is fair.

Packet timeouts are not intended to be regularly reached, just a fall-back for cases when no transactions can be confirmed due to full blocks or fee miscalculation. 

That said, if block time is always increasing (I forget whether Tendermint enforces this or not), it is also safe for use as a timeout marker. ICS 4 could be adapted to allow this option if it would be useful.

**Comment by mdyring** [comment]
It appears Tendermint does not currently enforce increasing block times. 

I've have opened a PR to solve this: https://github.com/tendermint/tendermint/pull/4353

To offer maximum flexibility, I think it would be a very valuable to support both block height and time-based timeouts.

**Comment by cwgoes** [comment]
I agree. It is not difficult to add this option to the specification; it could even be chosen per-packet.

**Comment by mdyring** [comment]
@melekes was kind enough to point out there is already a check for this. 

So it should be good to go on the Tendermint side:

https://github.com/tendermint/tendermint/blob/c5ecd802a22f82b0a4787c68d749fbc1da682306/state/validation.go#L100

**Comment by cwgoes** [comment]
Maybe we should just add this to `1.0.0`, it's really quite easy.

**Comment by mdyring** [comment]
> Maybe we should just add this to `1.0.0`, it's really quite easy.

Would be great, actually we were hoping to participate in Game of Zones with our em-ledger (based on Cosmos-SDK), so I guess we'd need something like this.

**Comment by cwgoes** [comment]
> > Maybe we should just add this to `1.0.0`, it's really quite easy.
> 
> Would be great, actually we were hoping to participate in Game of Zones with our em-ledger (based on Cosmos-SDK), so I guess we'd need something like this.

Aye, planning to land this before `1.0` at least.

**Comment by cwgoes** [comment]
@mdyring Would an option of *both* a block height timeout and a timestamp timeout be useful?

I think I might write it that way.

**Comment by mdyring** [comment]
For our use-case we're just interested in timestamps, but I guess(?) block heights could be as well.

So a packet will timeout if recipient chain has passed height X or time Y, right? If so, we'd just set X to be some very high value.

**Comment by cwgoes** [comment]
> For our use-case we're just interested in timestamps, but I guess(?) block heights could be as well.
> 
> So a packet will timeout if recipient chain has passed height X or time Y, right? If so, we'd just set X to be some very high value.

Yes, that's right, there could be a sentinel value of `0` to disable block height timeouts if you like.

**Comment by mdyring** [comment]
I think that would make sense, thanks!

**Comment by gjermundgaraba** [comment]
This issue was mistakenly reopened during some integration work, so closing again now.
